### PR TITLE
stablecoin V2

### PR DESCRIPTION
**Pull request by dorin-iancu**
Users can exchange their collateral for stablecoins and vice-versa. Contrary to the lending/borrowing style contracts, this SC acts as a simple exchange. Once collateral is deposited, the user does not own that collateral anymore.

Hedging agents can deposit some collateral to cover for the volatility. In more simple terms, they "bet" on collateral value increasing, in which case they receive a certain extra amount back on withdraw. If the value decreases, they lose part or all of their collateral.

Not done yet:
- pool rebalancing from hedging agent's deposits
- hedging position liquidation
- liquidity providers

### CODE REVIEW

File: stablecoin-v2/src/hedging_agents.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    fn open_hedging_position(
+        &self,
+        #[payment_token] payment_token: TokenIdentifier,
+        #[payment_amount] payment_amount: BigUint,
+        amount_to_cover: BigUint,
+        max_oracle_value: BigUint,
+    ) -> SCResult<()> {
+        self.require_collateral_in_whitelist(&payment_token)?;
+
+        let collateral_value_in_dollars = self.get_collateral_value_in_dollars(&payment_token)?;
```
Comment (0xmtf) [inline stablecoin-v2/src/hedging_agents.rs:35]: not sure if intended, but price decimals are not taken into consideration

File: stablecoin-v2/src/hedging_agents.rs [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+    // private
+
+    #[inline(always)]
+    fn calculate_leverage(
+        &self,
+        collateral_amount: &BigUint,
+        amount_to_cover: &BigUint,
+    ) -> BigUint {
+        self.calculate_ratio(&(collateral_amount + amount_to_cover), amount_to_cover)
```
Comment (0xmtf) [inline stablecoin-v2/src/hedging_agents.rs:143]: extract first arg as variable

File: stablecoin-v2/src/hedging_token.rs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            .esdt_system_sc_proxy()
+            .set_special_roles(
+                &own_sc_address,
+                &token_id,
+                (&roles[..]).into_iter().cloned(),
+            )
+            .async_call()
+    }
+
+    fn create_and_send_hedging_token(&self, to: &ManagedAddress) -> u64 {
```
Comment (0xmtf) [inline stablecoin-v2/src/hedging_token.rs:57]: i'd split the mint and transfer ops

File: stablecoin-v2/src/hedging_agents.rs (hunk 1 above)
Comment (dorin-iancu) [reply stablecoin-v2/src/hedging_agents.rs:35]: You're right, this calculation is wrong later down the line. Will fix in the next PR, as this part of the code is already slightly changed.

File: stablecoin-v2/src/hedging_agents.rs (hunk 2 above)
Comment (dorin-iancu) [reply stablecoin-v2/src/hedging_agents.rs:143]: I would, but I don't have a good name for it :/

**Review by 0xmtf — APPROVED** [review]

**Pull request merged**